lib/std/rand.zig

```diff
@@ -749,29 +749,31 @@ pub const Gimli = struct {
     random: Random,
     state: std.crypto.core.Gimli,
 
-    pub fn init(init_s: u64) Gimli {
+    pub const secret_seed_length = 32;
```

Comment:
This was explicitly named `secret_seed` instead of `seed` to emphasize the fact that unlike other RNGs, a secret seed is required here. With other RNGs, a simple counter can be totally fine.

16 bytes would still provide a 128-bit security level. But 32 bytes were chosen because:

- there is absolutely no setup overhead in having a 256 bit key vs a 128 bit key
- if the entropy source is biased, having more bits can only improve the min-entropy, avoid related keys, etc. This is especially an issue on embedded systems.
- protects against multi-target attacks, quantum attacks, etc. These may sound theoretical, but unlike AES256 vs AES128 that requires more rounds later, we can get this totally for free
- GIMLI-AEAD is specified for a 256 bit key only. libhydrogen uses 256 bit keys only. Essentially, all Gimli deployments today use a 256 bit key, so let's be boring rather than raise eyebrows about having made a different, less secure-looking choice :)

Reply:
> there is absolutely no setup overhead in having a 256 bit key vs a 128 bit key

The particular example I'm thinking of is that `AT_RANDOM` (the only "free" source of randomness) only gives us 16 bytes of randomness.

Reply:
The CSPRNG needs to be seeded only once, so it is okay for it not to be free.

And, unlike other system-provided entropy sources, `AT_RANDOM` is never cleared. If it gets leaked, all previous and future secrets can be recovered. 

And Gimli is a bit of a stopgap solution. Since efficiency of hardware implementations is irrelevant here, we may want to eventually switch to ChaCha, that will also require a 256 bit key (well... it can be used with a 128 bit key, but no one does that, and we don't implement that variant either).

Reply:
> Unlike other system-provided entropy sources, `AT_RANDOM` is never cleared.

If we were to start using it; we can also choose to clear it (which I believe glibc does?) or alternatively just replace it with some generated randomness.

Reply:
What if external libraries try to use it?

I haven't looked at glibc, but at least must doesn't seem to clear it.

Reply:
musl, not must. Damn you, autocorrect :)

Reply:
The first bytes of `AT_RANDOM` are used as stack canaries.  Reusing them for a different purpose doesn't make me very comfortable TBH. Leak one, get the other for free :(

Reply:
> The first bytes of `AT_RANDOM` are used as stack canaries.

That's a glibc thing IIRC. I was talking about only using `AT_RANDOM` in non-libc linked builds.

Reply:
Without a libc, if `getrandom()` is really not an option (not sure why... if `AT_RANDOM` exists, the kernel RNG has already been seeded and the `getrandom` syscall will never block), a 128 bit key can always be padded or duplicated.